rfd/0001-kubernetes-manager.md

```diff
@@ -0,0 +1,100 @@
+
+Then one can create a credentials secret with something like that:
+
+```bash
+tctl auth sign --user=kubernetes-manager --ttl=8760h --overwrite --format=file -o /tmp/kubernetes-manager-identity
+kubectl -n teleport create secret generic some-creds --from-file=identity=/tmp/kubernetes-manager-identity
+rm -f /tmp/kubernetes-manager-identity
+```
+
+Secret in Kubernetes is a key-value object, and the only supported key, for now, is `"identity"` that stores Teleport connection credentials.
+
+In the future, the process of identity derivation might be simplified by adding to `tctl auth sign` support of writing an identity directly to Kubernetes instead of a temporary file.
+
+### Resource definitions
+
+Custom resource definitions of Teleport resource look like this:
+
+```yaml
+apiVersion: resources.goteleport.com/v2
+kind: Role
+metadata:
+  name: folks
+  namespace: teleport
+instanceName: some-instance
+spec:
+  allow:
+    request:
+      roles: ["admin"]
+---
+apiVersion: resources.goteleport.com/v4
+kind: User
+metadata:
+  name: human-bean
+  namespace: teleport
+instanceName: some-instance
+spec:
+  roles: ["folks"]
+```
+
+Manager process watches for custom resource events. Once a custom resource is created/updated, the manager's watcher acknowledges it and upserts a resource into Teleport. Every custom resource must have the `instanceName` field useful to look up an instance resource in order to know what address to connect to. Instance resource and its associated secret must reside in the same namespace as the custom resource.
+
+The name of the Kubernetes custom resource is the name of the resource in Teleport. Also, every Kubernetes custom resource must have a `spec` field that has the same schema as a `spec` field of a resource in Teleport.
+
+Deletion of resources is being handled using [resource finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/). Every acknowledged resource gets a finalizer named `"resources.goteleport.com/delete"` so when the deletion happens, Kubernetes doesn't actually delete it but sets a deletion timestamp. While the finalizer list is non-empty, the manager can still access the resource's fields, `instanceName` in particular. Deletion of the Kubernetes object leads to deletion of the associated resource in Teleport. Once a deletion in Teleport succeeds, the manager removes a finalizer from the Kubernetes object so it can be garbage-collected.
+
+### API Groups and Versions
+
+`Instance` resource resides in an API group called `instances.goteleport.com`, and all the Teleport resources reside in a group called `resources.goteleport.com`. Two different groups are used for a reason:
+
+- API Versions of `instances.goteleport.com` start from `v7` because the Teleport version is 7.x at the moment of writing this document.
+- API Versions of `resources.goteleport.com` are all different because they're tied to Teleport resource versions.
```

Comment:
For simplicity, I'd probably just the same API group for both - but that's personal preference.